internal/indexer/controller/coalesce.go

```diff
@@ -13,22 +13,55 @@ import (
 
 // coalesce calls each ecosystem's coalescer and merges the returned IndexReports
 func coalesce(ctx context.Context, s *Controller) (State, error) {
-	coalescers := []indexer.Coalescer{}
+	cctx, cancel := context.WithCancel(ctx)
```

Comment:
Why this inner context? I don't think anything gets spawned outside the errgroup. Using `errgroup.WithContext` would be simpler

Reply:
If the main go routine encounters an error dispatching Coalesce go routines, I want the entire tree of routines to end, so I derive a ctx with cancel immediately and then defer the cancel. If the main go routines ends with an exception, all children are killed as well. 

it seems like errgroup.WithContext() just adds an additional ctx to maintain in the method and doesn't provide much other value. 

for example:

```
// coalesce calls each ecosystem's coalescer and merges the returned IndexReports
func coalesce(ctx context.Context, s *Controller) (State, error) {
	cctx, cancel := context.WithCancel(ctx)
	defer cancel() // still need to derive this so main routine can cancel
	mu := sync.Mutex{}
	reports := []*claircore.IndexReport{}
	g, ctx := errgroup.WithContext(cctx) // don't have much use for this derived ctx
	// dispatch a coalescer go routine for each ecosystem
	for _, ecosystem := range s.Ecosystems {
```

instead I just pass the launched go routines the cctx which, if there is an error in the dispatching main go routine, any inflight will cancel.

Reply:
Oh, I see, because the layer artifacts are being constructed in serial and then only concurrently coalescing at the end.

Reply:
yup, kinda like a "prep then dispatch" thing.